File: frappe_npo/overrides/sales_invoice.js
frappe.ui.form.on("Sales Invoice", {
	refresh: function (frm) {
		if (frm.doc.docstatus === 0 && !frm.doc.is_return) {
			frm.add_custom_button(
				__("Disbursement Order"),
				function () {
					frm.events.show_disbursement_order_selector(frm);
				},
				__("Get Items From"),
			);
		}
	},

	show_disbursement_order_selector: function (frm) {
		let d;

		const fetch_data = () => {
			frappe.call({
				method: "frappe_npo.overrides.sales_invoice.get_disbursement_orders",
				args: {
					from_date: d.get_value("from_date"),
					to_date: d.get_value("to_date"),
					company: d.get_value("company"),
					donor: d.get_value("donor"),
					po_no: d.get_value("po_no"),
					disbursement_type: d.get_value("disbursement_type"),
				},
				callback: function (r) {
					let grid = d.fields_dict.orders_table.grid;
					grid.df.data = [];

					if (r.message) {
						r.message.forEach((row) => {
							grid.df.data.push({
								selected: 0,
								name: row.name,
								donor: row.donor,
								po_no: row.po_no,
								po_date: row.po_date,
							});
						});
					}

					grid.refresh();
				},
			});
		};

		const debounce_fetch = frappe.utils.debounce(fetch_data, 300);

		d = new frappe.ui.Dialog({
			title: __("Select Disbursement Orders"),
			size: "large",
			fields: [
				{
					label: __("Donor"),
					fieldname: "donor",
					fieldtype: "Link",
					options: "Donor",
					onchange: debounce_fetch,
					reqd: 1,
				},
				{
					label: __("From Date"),
					fieldname: "from_date",
					fieldtype: "Date",
					onchange: debounce_fetch,
				},
				{
					label: __("To Date"),
					fieldname: "to_date",
					fieldtype: "Date",
					onchange: debounce_fetch,
				},
				{ fieldtype: "Column Break" },
				{
					label: __("Company"),
					fieldname: "company",
					fieldtype: "Link",
					options: "Company",
					default: frm.doc.company,
					onchange: debounce_fetch,
				},
				{
					label: __("Disbursement Type"),
					fieldname: "disbursement_type",
					fieldtype: "Select",
					options: "\nCash\nPhysical Goods\nServices",
					onchange: debounce_fetch,
				},
				{
					label: __("PO No"),
					fieldname: "po_no",
					fieldtype: "Data",
					onchange: debounce_fetch,
				},
				{ fieldtype: "Section Break" },
				{
					label: __("Disbursement Orders"),
					fieldname: "orders_table",
					fieldtype: "Table",
					cannot_add_rows: true,
					in_place_edit: false,
					fields: [
						{
							fieldtype: "Link",
							fieldname: "name",
							options: "Disbursement Order",
							label: __("Order ID"),
							in_list_view: 1,
							read_only: 1,
						},
						{
							fieldtype: "Link",
							fieldname: "donor",
							options: "Donor",
							label: __("Donor"),
							in_list_view: 1,
							read_only: 1,
						},
						{
							fieldtype: "Data",
							fieldname: "po_no",
							label: __("PO No"),
							in_list_view: 1,
							read_only: 1,
						},
						{
							fieldtype: "Date",
							fieldname: "po_date",
							label: __("PO Date"),
							in_list_view: 1,
							read_only: 1,
						},
					],
				},
			],
			primary_action_label: __("Fetch Items"),
			primary_action: function () {
				const selected_rows = d.fields_dict.orders_table.grid.get_selected_children();

				if (!selected_rows.length) {
					frappe.msgprint(__("Please select at least one order"));
					return;
				}

				frappe.call({
					method: "frappe_npo.overrides.sales_invoice.get_disbursement_totals",
					args: { orders: selected_rows.map((r) => r.name) },
					callback: async function (r) {
						if (r.message) {
							const { items, customer, donor } = r.message;

							if (customer) frm.set_value("customer", customer);
							if (donor) frm.set_value("donor", donor);
							if (items.length) {
								frm.set_value("disbursement_order", items[0].disbursement_order);
							}

							frm.clear_table("items");

							const company_resp = await frappe.db.get_value(
								"Company",
								frm.doc.company,
								"default_income_account",
							);
							const company_default = company_resp.message.default_income_account;

							items.forEach((item) => {
								let row = frm.add_child("items");

								row.item_code = item.item_code;
								row.item_name = item.item_name;
								row.uom = item.uom;
								row.project = item.project;
								row.qty = item.qty || 1;
								row.rate = item.rate;
								row.amount = item.amount;
								row.disbursement_order = item.disbursement_order;

								if (item.description) {
									row.description = item.description;
								}
								if (!row.income_account && company_default) {
									row.income_account = company_default;
								}
							});

							frm.refresh_field("items");
							frm.trigger("calculate_taxes_and_totals");

							d.hide();
						}
					},
				});

				d.hide();
			},
		});

		d.show();
	},
});

File: frappe_npo/overrides/sales_invoice.py
import frappe
from frappe import _
import json
from frappe.query_builder.functions import Sum


@frappe.whitelist()
def get_disbursement_orders(
    from_date=None,
    to_date=None,
    company=None,
    donor=None,
    po_no=None,
    disbursement_type=None,
    item_code=None,
):
    filters = {"docstatus": 1}

    if from_date and to_date:
        filters["creation"] = ["between", [from_date, to_date]]
    elif from_date:
        filters["creation"] = [">=", from_date]
    elif to_date:
        filters["creation"] = ["<=", to_date]

    if company:
        filters["company"] = company
    if donor:
        filters["donor"] = donor
    if po_no:
        filters["po_no"] = ["like", f"%{po_no}%"]
    if disbursement_type:
        filters["disbursement_type"] = disbursement_type

    disbursement_orders = frappe.get_all(
        "Disbursement Order",
        filters=filters,
        fields=[
            "name",
            "donor",
            "po_no",
            "po_date",
            "disbursement_type",
        ],
        order_by="creation desc",
    )

    unbilled_orders = []

    for order in disbursement_orders:
        existing_invoice = frappe.db.exists(
            "Sales Invoice Item",
            {
                "disbursement_order": order.name,
                "docstatus": ["!=", 2],
            },
        )

        if not existing_invoice:
            unbilled_orders.append(order)

    return unbilled_orders


@frappe.whitelist()
def get_disbursement_totals(orders):
    if isinstance(orders, str):
        orders = json.loads(orders)

    if not orders:
        return {}

    do = frappe.qb.DocType("Disbursement Order")
    donor_dt = frappe.qb.DocType("Donor")

    header_data = (
        frappe.qb.from_(do)
        .join(donor_dt)
        .on(do.donor == donor_dt.name)
        .select(do.donor, donor_dt.customer)
        .where(do.name.isin(orders))
    ).run(as_dict=True)

    unique_donors = list(set(d.donor for d in header_data))
    unique_customers = list(set(d.customer for d in header_data))

    customer = unique_customers[0] if len(unique_customers) == 1 else None
    donor = unique_donors[0] if len(unique_donors) == 1 else None

    pe = frappe.qb.DocType("Payment Entry")
    payments = (
        frappe.qb.from_(pe)
        .select(pe.disbursement_order, Sum(pe.paid_amount).as_("total_paid"))
        .where(pe.disbursement_order.isin(orders))
        .where(pe.payment_type == "Pay")
        .where(pe.docstatus == 1)
        .groupby(pe.disbursement_order)
    ).run(as_dict=True)

    paid_map = {p.disbursement_order: p.total_paid for p in payments}

    dop = frappe.qb.DocType("Disbursement Order Party")
    fees = (
        frappe.qb.from_(dop)
        .select(dop.parent, Sum(dop.bank_transfer_fee).as_("total_fee"))
        .where(dop.parent.isin(orders))
        .groupby(dop.parent)
    ).run(as_dict=True)

    fee_map = {f.parent: f.total_fee for f in fees}

    items_list = []

    for order_name in orders:
        doi = frappe.qb.DocType("Disbursement Order Item")
        order_items = (
            frappe.qb.from_(doi)
            .select(
                doi.item_code, doi.project, doi.uom, doi.item_name, doi.rate, doi.qty
            )
            .where(doi.parent == order_name)
        ).run(as_dict=True)

        order_payment = paid_map.get(order_name, 0)
        order_fee = fee_map.get(order_name, 0)

        for i, item in enumerate(order_items):
            if i == 0:
                calc_rate = (order_payment or 0) + (order_fee or 0)
                calc_qty = 1
            else:
                calc_rate = item.rate
                calc_qty = item.qty

            items_list.append(
                {
                    "item_code": item.item_code,
                    "item_name": item.item_name or item.item_code,
                    "qty": calc_qty,
                    "uom": item.uom,
                    "rate": calc_rate,
                    "amount": float(calc_rate) * float(calc_qty),
                    "project": item.project,
                    "disbursement_order": order_name,
                }
            )

    return {"items": items_list, "customer": customer, "donor": donor}
